### ISSUE TITLE
Analytics - Difference Between Total Results and Actual Number of Rows

### ISSUE THREAD

**Issue by jorgemd24** [woocommerce/woocommerce#49635]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

While working on this PR [https://github.com/woocommerce/woocommerce/pull/49562](<https://github.com/woocommerce/woocommerce/pull/49562>), I realized that sometimes the total displayed in the table doesn't match the actual number of results. See the image here:

![image](https://uploads.linear.app/f629fcde-6f9b-48e8-8af3-60eed8823bff/682780a3-d501-4aa8-9984-8a7def3bafb4/e526ca8e-7adf-4010-9829-2f819186e05e?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiL2Y2MjlmY2RlLTZmOWItNDhlOC04YWYzLTYwZWVkODgyM2JmZi82ODI3ODBhMy1kNTAxLTRhYTgtOTk4NC04YTdkZWYzYmFmYjQvZTUyNmNhOGUtN2FkZi00MDEwLTk4MjktMmY4MTkxODZlMDVlIiwiaWF0IjoxNzg0MDg0MjA1LCJleHAiOjE4MTU2NTQ3NjV9.LrFRY0JE4rTmeQGkEl9HiGJkP06ATP7SrC0raphdp8E)

This bug might seem minor, but it causes an issue because the endpoint to calculate the total number of results and the data are also used to create the CSV reports.

[https://github.com/woocommerce/woocommerce/blob/db9ef8b5775eb3c08af3b8ae00b9c6b60106251c/plugins/woocommerce/src/Admin/ReportCSVExporter.php#L238-L241](<https://github.com/woocommerce/woocommerce/blob/db9ef8b5775eb3c08af3b8ae00b9c6b60106251c/plugins/woocommerce/src/Admin/ReportCSVExporter.php#L238-L241>)

The code expects the percentage to be 100% to email the CSV report, so if these two parameters don't match, the CSV reports will never be sent.

[https://github.com/woocommerce/woocommerce/blob/db9ef8b5775eb3c08af3b8ae00b9c6b60106251c/plugins/woocommerce/src/Admin/ReportExporter.php#L202-L212](<https://github.com/woocommerce/woocommerce/blob/db9ef8b5775eb3c08af3b8ae00b9c6b60106251c/plugins/woocommerce/src/Admin/ReportExporter.php#L202-L212>)

For example, in the Tax Report, the affected queries are this ones:

[https://github.com/woocommerce/woocommerce/blob/db9ef8b5775eb3c08af3b8ae00b9c6b60106251c/plugins/woocommerce/src/Admin/API/Reports/Taxes/DataStore.php#L187-L204](<https://github.com/woocommerce/woocommerce/blob/db9ef8b5775eb3c08af3b8ae00b9c6b60106251c/plugins/woocommerce/src/Admin/API/Reports/Taxes/DataStore.php#L187-L204>)

### Expected behavior

The total results and the retrieved data should be consistent.

### Actual behavior

The total results and the retrieved data might show different totals.

### Steps to reproduce

I don't have specific steps because I'm unsure of the exact reason they don't match, but I'm investigating to find the difference.

### WordPress Environment

WP 6.5<br>WC 9.1.2

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](<https://woocommerce.com/products/storefront/>).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by tomalec** [comment]
I think the issue may be related to faulty `per_page` handling. as the request with 
`/wp-json/wc-analytics/reports/taxes?orderby=tax_rate_id&order=desc&after=2024-07-01T00%3A00%3A00&before=2024-07-19T23%3A59%3A59&page=1&per_page=25&_locale=user`
```
orderby: tax_rate_id
order: desc
after: 2024-07-01T00:00:00
before: 2024-07-19T23:59:59
page: 1
per_page: 25
_locale: user
```
returns 33 results for me


Also marking as potentially related https://github.com/woocommerce/woocommerce/pull/46248

**Comment by tomalec** [comment]
I think it may be related to the fact the records are counted before grouping the results 

https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Admin/API/Reports/Taxes/DataStore.php#L189-L206


Also marking as potentially related https://github.com/woocommerce/woocommerce/pull/46248

**Comment by tomalec** [comment]
I was unable to replicate the issue consistently. I was able to see that in July, but not today.

However, from the code investigation, I found a potential reason of the rendered values mismatch:
`<TaxesReportTable>` component reads data from two separate endpoints: `/reports/taxes/` and `/reports/taxes/stats/`. It uses `/reports/taxes/` to list the taxes in the table, but `/reports/taxes/stats` to fetch the total number of taxes for the pagination summary.
https://github.com/woocommerce/woocommerce/blob/43cc912bbaa75fb697c4acf5bc6f6b317339762f/plugins/woocommerce-admin/client/analytics/report/taxes/table.js#L136-L147

So, I guess the issue may happen when the two responses get out of sync, possibly due to caching or data changes in the race condition in between the requests.

The solution for the wrong number rendered would be to use `totalResults` number provided by `<Table>` component to the `getSummary` function, which is exactly the number of items from the `/reports/{endpoint}`. Like https://github.com/woocommerce/woocommerce/pull/51218 However, this may hide an issue lying somewhere deeper

:warning: In case of a response mismatch, other numbers, like "total tax, order tax, shipping tax," may also be incorrect.

Plus, it still does not answer the export problem.

**Comment by tomalec** [comment]
Other related issues:

- https://github.com/woocommerce/woocommerce/issues/48841 fixed by https://github.com/woocommerce/woocommerce/pull/49283
- https://github.com/woocommerce/woocommerce/issues/38618 - The tax list is empty, but the chart has some data.
- https://github.com/woocommerce/woocommerce/issues/38347
- https://github.com/woocommerce/woocommerce/issues/38612

**Comment by tomalec** [comment]
`/reports/taxes/` passes ids [through filters via `get_filtered_ids`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Admin/API/Reports/Taxes/DataStore.php#L141:L141) while [`/reports/taxes/stats` does not](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Admin/API/Reports/Taxes/Stats/DataStore.php#L91-L95)  
That could be a cause of the inconsistency, if the filter was used

Here is the PR with a fix https://github.com/woocommerce/woocommerce/pull/51219.

However, I guess in OP you were not using `woocommerce_analytics_taxes` filter. Also, I didn't find any extension in `all-plugins` that uses it.
So, I keep on investigating.

**Comment by tomalec** [comment]
I found that the actual steps to reproduce the visual mismatch in https://github.com/woocommerce/woocommerce/issues/48841

1. Place an order with taxes.
2. Refund only the taxes of the order.
3. Check the totals in the taxes report (attached image)

It's due to the fact the total number in the table summary is fetched from `/stats` request

**Comment by tomalec** [comment]
Another case to make results in just `/reports/taxes/` inconsistent, and break the CSV export:

1. Import some taxes
2. Create a paid order with tax, let's say `PL-TAX-PL-1, 18%`
3. Go to `/wp-admin/admin.php?page=wc-settings&tab=tax&section=standard` and change the value of that tax `PL-TAX-PL-1, 21%`
4. Create a paid order with that tax
5. Go to `/wp-admin/admin.php?page=wc-admin&path=%2Fanalytics%2Ftaxes`
6. Select a period that covers the order and the refund ( and preferably large number of other taxes for export)
7. Go to the last page of the table
8. Check the network tab in dev tools. Check `taxes` request. `x-wp-total:` header do not match the number of results in the response
![image](https://github.com/user-attachments/assets/699fa626-67de-45d0-bb5b-cf4216549ba4)

9. Click "Download"
11. Wait for the export action to be scheduled
12. Check that no email was sent


As stated https://github.com/woocommerce/woocommerce/issues/49635#issuecomment-2238560754 and the separate issue, the group by is the reason

**Comment by tomalec** [comment]
@martynmjones I managed to find consistently reproducible steps. Updated the OP

**Comment by louwie17** [comment]
@tomalec I also just ran across this issue: https://wordpress.org/support/topic/tax-report-still-everything-duplicated/ which is related to the taxes.
It appears something weird happens when you have an order that includes products with multiple different tax classes ( see **Additional tax classes** setting to add additional ones ).

**Comment by tyxla** [comment]
I can no longer reproduce this one. I believe it was fixed by #51270.

Closing, feel free to reopen with specific instructions if it happens again.
